[clang][frontend] Add support for attribute plugins for statement attributes

ericastor opened the pull request:
We already have support for declaration attributes; this is just a matter of extending the plugin infrastructure to cover one more case.

ericastor committed:
[clang][frontend] Add support for attribute plugins for statement attributes

We already have support for declaration attributes; this is just a matter of extending the plugin infrastructure to cover one more case.

ericastor committed:
Merge branch 'main' into stmt-attributes

ericastor committed:
Fix desync'd comment

ilya-biryukov commented:
This looks like a relatively straightforward extension in line with the current Clang interfaces. Therefore, I feel we should accept this.

However, please wait for feedback from Clang owners to make sure this does not go against some non-obvious trade-offs or future plans that Clang has.

cc @AaronBallman, @cor3ntin

Sirraide reviewed (commented):
Looks ok to me, but I’ll leave it to @erichkeane to approve this as he’s the attributes code owner.

erichkeane reviewed (commented):
So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.

So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).

ericastor commented:
> So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.
> 
> So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).

This is a fair point... how are standard statement attributes handled inside of templates? (Assuming clang supports that.)

erichkeane commented:
> > So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.
> > So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).
> 
> This is a fair point... how are standard statement attributes handled inside of templates? (Assuming clang supports that.)

All our handling of attributes is pretty manual for anything with arguments.  We have a bit of an automatic handling for many attributes, but arguments aren't.  I think there is a section in `SemaInstantiateDecl.cpp` that handles a lot of it (`InstantiateAttrs`) and you can see there is a LOT of manual handling there

ericastor committed:
Add support for template-instantiations of AnnotateAttr

ericastor commented:
So... I think this turned out to be surprisingly easy. After all, we don't actually allow custom attributes to create entirely new attributes in the AST, since they can't create new legal AST entries... instead, the example creates AnnotateAttrs that can then be observed & acted on by the compiler-extending plugin.

It looks like reaching the same level of support just required adding AnnotateAttr support to the template instantiation logic. Am I missing anything else? @erichkeane

erichkeane commented on clang/lib/Sema/SemaTemplateInstantiate.cpp:
```diff
@@ -2125,6 +2126,19 @@ TemplateInstantiator::TransformTemplateParmRefExpr(DeclRefExpr *E,
+      return AA;
+    Args.push_back(Res.get());
+  }
+  return AnnotateAttr::CreateImplicit(getSema().Context, AA->getAnnotation(),
```
we probably want to transform the annotation as well?  or does `StringArgument` not work for dependent values?

ericastor replied:
I think `StringArgument` only handles `StringRef`s? I don't see a way it could be given a dependent value.

erichkeane reviewed (commented):
This seems reasonable?  I actually didn't know we weren't letting them create new attributes, and were just using `AnnotateAttr` in the AST.  I think @AaronBallman should do 1 run through this as he understands the plugins better than I, but this otherwise LGTM.

ericastor committed:
Merge branch 'main' into stmt-attributes

ericastor committed:
Merge branch 'llvm:main' into stmt-attributes

jhuber6 committed:
[libc] Fix missing namespace declarations

ericastor committed:
[clang][frontend] Support applying the annotate attribute to statements (#111841)

By allowing AnnotateAttr to be applied to statements, users can place arbitrary information in the AST for later use.

For example, this can be used for HW-targeted language extensions that involve specialized loop annotations.

jhuber6 committed:
[libc] Add missing config include

JDevlieghere committed:
[lldb] Fix a variety of LLDB_LOG format strings

LLVM now triggers an assertion when the format string and arguments
don't match. Fix a variety of incorrect format strings I discovered when
enabling logging with a debug build.

tatwaichong committed:
[mlir][tosa] Change the type of profile option to ListOption (#111214)

In tosa valiation pass, change the type of profile option to ListOption.
Now TOSA profiles is turned from hierarchical to composable. Each
profile is an independent set, i.e. an target can implement multiple
profiles.

Set the profile option to none by default, and limit to profiles if
requested.
The profiles can be specified via command line, e.g.
$ mlir-opt ... --tosa-validate="profile=bi,mi" which tells the valiation
pass that BI and MI are enabled.

Change-Id: I1fb8d0c1b27eccd768349b6eb4234093313efb57

rniwa committed:
[alpha.webkit.UncountedCallArgsChecker] Skip std::forward in tryToFindPtrOrigin. (#111222)

Ignore std::forward when it appears while looking for the pointer
origin.

rniwa committed:
[alpha.webkit.UncountedCallArgsChecker] Add the support for trivial CXXInheritedCtorInitExpr. (#111198)

rniwa committed:
isUncountedPtr should take QualType as an argument. (#110213)

Make isUncountedPtr take QualType as an argument instead of Type*. This
simplifies some code.

abidh committed:
[mlir][debug] Support DICommonBlock. (#111706)

A COMMON block is a named area of memory that holds a collection of
variables. Fortran subprograms may map the COMMON block memory area to a
list of variables. A common block is represented in LLVM debug by
DICommonBlock.

This PR adds support for this in MLIR. The changes are mostly mechanical
apart from small change to access the DICompileUnit when the scope of
the variable is DICommonBlock.

---------

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

ldionne committed:
[runtimes][NFC] Reindent CMake files (#111821)

This is a purely mechanical commit for fixing the indentation of the
runtimes' CMakeLists files after #80007. That PR didn't update the
indentation in order to make the diff easier to review and for merge
conflicts to be easier to resolve (for downstream changes).

This doesn't change any code, it only reindents it.

justinfargnoli committed:
[NVPTX] Prefer prmt.b32 over bfi.b32 (#110766)

In [[NVPTX] Improve lowering of
v4i8](https://github.com/llvm/llvm-project/commit/cbafb6f2f5c99474164dcc725820cbbeb2e02e14)
@Artem-B add the ability to lower ISD::BUILD_VECTOR with bfi PTX
instructions. @Artem-B did this because:
([source](https://github.com/llvm/llvm-project/pull/67866#discussion_r1343066911))

> Under the hood byte extraction/insertion ends up as BFI/BFE
instructions, so we may as well do that in PTX, too.
https://godbolt.org/z/Tb3zWbj9b

However, the example that @Artem-B linked was targeting sm_52. On modern
architectures, ptxas uses prmt.b32.
[Example](https://godbolt.org/z/Ye4W1n84o).

Thus, remove uses of NVPTXISD::BFI in favor of NVPTXISD::PRMT.

klausler committed:
[flang][runtime] Fix runtime crash after bad recoverable OPEN (#111454)

When an OPEN statement with a unit number fails in a recoverable manner,
the runtime needs to delete the ExternalFileUnit instance that was
created in the unit map. And we do this too soon -- that instance still
holds some of the I/O statement state that will be used by a later call
into the runtime for EndIoStatement.

Move the code that deletes the unit after a failed but recoverable OPEN
into ExternalIoStatementBase::EndIoStatement, and don't do things
afterwards that would need the I/O statement state that has been
destroyed.

Fixes https://github.com/llvm/llvm-project/issues/111404.

klausler committed:
[flang] Fix references to destroyed objects (#111582)

ProgramTree instances are created as the value of a local variable in
the Pre(const parser::ProgramUnit &) member function in name resolution.
But references to these ProgramTree instances can persist in
SubprogramNameDetails symbol table entries that might survive that
function call's lifetime, and lead to trouble later when (e.g.)
expression semantics needs to deal with a possible forward reference in
a function reference in an expression being processed later in
expression checking.

So put those ProgramTree instances into a longer-lived linked list
within the SemanticsContext.

Might fix some weird crashes reported on big-endian targets (AIX &
Solaris).

klausler committed:
[flang] Minor cleanup (move function into /tools.cpp) (#111587)

The semantics utility GetAllNames has declarations in two header files
and a definition that really should be in the common utilities source
file. Remove the redudant declaration from resolve-names-utils.h and
move code from resolve-names-utils.cpp into Semantics/tools.cpp.

vogelsgesang committed:
[lldb][libc++] Hide all libc++ implementation details from stacktraces (#108870)

This commit changes the libc++ frame recognizer to hide implementation
details of libc++ more aggressively. The applied heuristic is rather
straightforward: We consider every function name starting with `__` as
an implementation detail.

This works pretty neatly for `std::invoke`, `std::function`,
`std::sort`, `std::map::emplace` and many others. Also, this should
align quite nicely with libc++'s general coding convention of using the
`__` for their implementation details, thereby keeping the future
maintenance effort low.

However, this heuristic by itself does not work in 100% of the cases:
E.g., `std::ranges::sort` is not a function, but an object with an
overloaded `operator()`, which means that there is no actual call
`std::ranges::sort` in the call stack. Instead, there is a
`std::ranges::__sort::operator()` call. To make sure that we don't hide
this stack frame, we never hide the frame which represents the entry
point from user code into libc++ code

clementval committed:
[flang][runtime][NFC] Fix header guard typo (#111741)

Header guard was in sync with the filename.

rengolin committed:
Fix GCC build problem with 03483737a7a2

ericastor committed:
Merge branch 'main' into stmt-attributes

AaronBallman reviewed (approved):
LGTM, though we should probably add a release note to clang/docs/ReleaseNotes.rst so users know that the plugin system got a bit more functionality.

ericastor committed:
Add release note

ericastor committed:
Merge branch 'main' into stmt-attributes
